### DISCUSSION TITLE
Linguist concludes a JS/markdown repo with a single perl file is somehow "46.2% Perl"

### DISCUSSION THREAD

**Discussion by Pomax** [github-linguist/linguist#5455] [Classification] [answered]
https://github.com/Pomax/BezierInfo-2 consists of 117 javascript files, 93 markdown files, 21 html files, 3 php files, 2 css files, and a single perl file. Yet somehow, the breakdown is:

Perl 46.2%
JavaScript 42.4%
HTML 9.7%
PHP 1.7%

Rather than the expected:

JavaScript 49.3%
Markdown 39.2%
HTML 8.8%
PHP 1.2%

so something seems to be going very wrong here =)

**Comment by lildude** [comment, marked as answer]
> so something seems to be going very wrong here =)

Not from a Linguist perspective 😉 Maybe you missed the following when you read [How Linguist Works](https://github.com/github/linguist/blob/master/docs/how-linguist-works.md):

> Linguist starts by going through all the files in a repository and excludes all files that it determines to be binary data, [vendored code](/docs/overrides.md#vendored-code), [generated code](/docs/overrides.md#generated-code), [documentation](/docs/overrides.md#documentation), or are defined as `data` (e.g. SQL) or `prose` (e.g. Markdown) languages, whilst taking into account any [overrides](/docs/overrides.md)

This means all files in your `docs` dir will be ignored by default as they're considered documentation and all Markdown files will be ignored because they're considered `prose`. 

Thus leaving only the files in `src` which contains your Perl file and brings us to the next point in that document:

> The percentages are calculated based on the bytes of code for each language as reported by the [List Languages](https://docs.github.com/rest/reference/repos#list-repository-languages) API.

That single Perl file contains more bytes of code than all your JavaScript combined:

```console
$ curl -H "Accept: application/vnd.github.v3+json" https://api.github.com/repos/Pomax/BezierInfo-2/languages
{
  "Perl": 62200,
  "JavaScript": 57149,
  "HTML": 13045,
  "PHP": 2247
}
```

So all in, Linguist is working as expected. If you wish to change this behaviour, you'll need to implement the relevant overrides to suit your desires.

**Reply by Pomax** [reply]
Hold up. Now we're passing the buck: Github uses Linguist to tell people which languages repos break down as. Github also _forces_ people to use a `docs` dir if they want gh-pages "not at the root level": you only get two choices. So yeah: I'd love to change that dir name, I hate having to use `docs` because it's not documentation, it's source, I've filed feature requests to add "picking a dir in a repo" to the gh pages concept and nothing has moved on that (hopefully "yet") but as it stands, neither I nor anyone else has agency here because Github has not granted people that power.

The point of the language breakdown is to give users an indication of a project breakdown, which I assume Github wants to get right, so I'm reporting that Linguist is doing the wrong thing for the role it's in on Github, even if as per the current code it's doing exactly what it was written to do. What it was written to do turns out to end up not matching what its role on github.com is: surely that's worth looking into.

**Reply by lildude** [reply]
There's no buck-passing at all and Linguist isn't doing anything wrong. Most of the decisions around how Linguist behaves are to cater for the largest possible audience and not get it 100% right, 100% of the time; this would be impossible due to personal preference alone, which is the case now.

A vast majority of people, whether they use Pages or not, use the `docs` dir for documentation, and the content is mostly prose, whether that's in text or Markdown or any other documentation format, so most users don't want prose counted and displayed as one of their programming languages. That's what Linguist caters for.

> neither I nor anyone else has agency here because Github has not granted people that power.

Not in the context of Pages, but you have loads of agency WRT Linguist. Your preferred behaviour doesn't match the wider usage Linguist is catering for and as such Linguist provides you with the option of implementing an override so it will adhere to your desired outcome. So whilst you can't change the directory name, you can tell Linguist to not treat it as documentation. You can also tell it to count your Markdown files rather than ignore them.

If you add a `.gitattributes` file with this in it:

```gitattributes
*.md linguist-detectable
docs/** -linguist-documentation
```

... you'll get the breakdown you want:

![CleanShot 2021-07-12 at 17 37 12](https://user-images.githubusercontent.com/627280/125324508-e0d9ab00-e337-11eb-8a04-cdc76c6b9e04.png)

You'll need to make another change to another file to force a reanalysis as adding the `.gitattributes` file isn't sufficient to force a reanalysis.

**Reply by Pomax** [reply]
Ah, fantastic: that's the missing piece.

Can that please be added as short section to the README.md so that the fact that that file exists is immediately discoverable when folks follow the link that github supports gives them? It doesn't need to be a full explanation, just something along the lines of "if you need to change how linguist parses your repository, you can do this using a `.gitattributes` file, click [here](...) for detailed information"?

Maybe even under the "Usage" heading, since it's basically using linguist by proxy.

**Reply by lildude** [reply]
🤔 we do this already, though not in quite so many words, with links to the documentation right at the top of the README with the second link being pretty clear to me.

**Reply by Pomax** [reply]
hm, I don't see any mention of `.gitattributes` here? https://github.com/github/linguist/blob/master/README.md

The links to "the docs" are clear enough, but this is a classic case of "you don't know you even want to look for this in the docs unless you already know that's something you'll find in the docs": mentioning that Github users (rather than Linguist users) can configure how linguist works through a `.gitattribtues` file, directly in the README.md (without saying how, of course, just at least surfacing that as a thing they can do with links to the relevant section(s) in the docs), would have very high value, in addition to the "Application usage" and "Command line usage" sections.

**Reply by lildude** [reply]
We used to have all the documentation in the README however I moved it because it was cumbersome to maintain and not consistent with how you'd document the basic installation and usage of a Ruby library.

I'm apprehensive about re-introducing any mention of overrides etc back into the README as it will likely lead to further requests to "just add this" and soon we'll be back to square one. If anything is going to be documented for GitHub users (keep in mind GitHub isn't the only user of Linguist), it should be in the GitHub docs; after all, most GitHub users wouldn't know to come looking at Linguist unless referred here by support or the docs.

I admit the language coverage/config/stats info in the GitHub docs is lacking. I'm not responsible for those docs but can certainly bring it up with our docs team.

**Reply by Pomax** [reply]
I didn't know, so I contacted support, and they sent me here. So this entire conversation is evidence that it'd be good to at least mention there's a config file people can create. None of this conversation would have happened if that'd been at least mentioned in the README =)
